module/zfs/zio.c

```diff
@@ -1626,19 +1615,33 @@ zio_vdev_delegated_io(vdev_t *vd, uint64_t offset, abd_t *data, uint64_t size,
 	return (zio);
 }
 
+
+/*
+ * Send a flush command to the given vdev. Unlike most zio creation functions,
+ * the flush zios are issued immediately. You can wait on pio to pause until
+ * the flushes complete.
+ */
 void
 zio_flush(zio_t *pio, vdev_t *vd)
 {
+	const zio_flag_t flags = ZIO_FLAG_CANFAIL | ZIO_FLAG_DONT_PROPAGATE |
+	    ZIO_FLAG_DONT_RETRY;
+	zio_t *zio;
+
 	if (vd->vdev_nowritecache)
 		return;
+
 	if (vd->vdev_children == 0) {
-		zio_nowait(zio_ioctl(pio, vd->vdev_spa, vd,
-		    DKIOCFLUSHWRITECACHE, NULL, NULL, ZIO_FLAG_CANFAIL |
-		    ZIO_FLAG_DONT_PROPAGATE | ZIO_FLAG_DONT_RETRY));
+		zio = zio_create(pio, pio->io_spa, 0, NULL, NULL, 0, 0, NULL,
+		    NULL, ZIO_TYPE_FLUSH, ZIO_PRIORITY_NOW, flags, vd, 0, NULL,
+		    ZIO_STAGE_OPEN, ZIO_FLUSH_PIPELINE);
 	} else {
+		zio = zio_null(pio, pio->io_spa, NULL, NULL, NULL, flags);
```

Comment:
I intentionally removed intermediate ZIO in 5a3bffab10799c.  Why are you resurrecting it?  What's wrong with assigning all children to the parent, if one provided?  NULL ZIOs are not free.

Also, by using pio->io_spa you've made pio a required argument in addition to vdev. It may not be a problem now, but what's the point?

Reply:
Ahh, I'd missed that it was removed on purpose. No problem then.